### PR TITLE
TTI

### PR DESCRIPTION
**Pull request by navjotk**
I am opening this PR to be able to discuss the status and progress of this branch. 
This isn't ready for merge yet.

### CODE REVIEW

**Comment by navjotk** [comment]
Current known problems: 
 - somehow the dampening term is missing from the generated code. Need to check why. 
 - There is no implementation for the trigonometric functions included in the kernel
 - The code generates a Seg Fault as it stands now. I haven't checked why but it's very likely to be related to either of the above problems.

**Comment by navjotk** [comment]
In principle, this version works without any compilation or runtime errors. I can't guarantee the correctness of the solution without a numerical test though.
For now, the trigonometric functions from math.h are being used.

File: devito/propagator.py [hunk 1]
```diff
@@ -105,6 +105,7 @@ def convert_equality_to_cgen(self, equality):
         try:
             return cgen.Assign(ccode(self.time_substitutions(equality.lhs).xreplace(self._var_map)), ccode(self.time_substitutions(equality.rhs).xreplace(self._var_map)))
         except:
+            raise
```
Comment (mlange05) [inline devito/propagator.py]: Not sure the `raise` here is intentional. Either raise a `RuntimeError` with an informative message, or don't catch the error at all.

File: examples/fwi_operators.py [hunk 2]
```diff
@@ -1,9 +1,135 @@
...
+        reference_cell = [(x1, 0),
+                          (y1, 0),
+                          (z1, 0),
+                          (x2, self.h),
+                          (y2, self.h),
+                          (z2, self.h)]
+        A = A.subs(reference_cell)
+        self.bs = A.inv().T.dot(p)
+
+    def point2grid(self, pt_coords):
```
Comment (mlange05) [inline examples/fwi_operators.py]: A proper docstring would be great here.

File: examples/tti_operators.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        B = delta(x, y, z)  # sqrt(1 + 2epsilon) but make the symbolic simpler
+        Th = theta(x, y, z)
+        Ph = phi(x, y, z)
+        # Weights to sum the two fields
+        # w1 = .5
+        # w2 = .5
+        dttp = as_finite_diff(p(x, y, z, t).diff(t, t), [t-s, t, t+s])
+        dttr = as_finite_diff(r(x, y, z, t).diff(t, t), [t-s, t, t+s])
+        # dtp = as_finite_diff(p(x, y, z, t).diff(t), [t-s, t])
+        # dtr = as_finite_diff(r(x, y, z, t).diff(t), [t-s, t])
```
Comment (mlange05) [inline examples/tti_operators.py]: These comments look like debug/dev leftovers. Are they necessary? Same goes for the symbol and parameter definitions above.

File: examples/tti_operators.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    def read(self, u, v):
+        eqs = []
+        for i in range(self.npoints):
+            eqs.append(Eq(self[t, i], (self.grid2point(u, self.orig_data[i, :]) + self.grid2point(v, self.orig_data[i, :]))/2))
+        return eqs
+
+
+class TTIOperator(Operator):
+    def _init_taylor(self, dim=2, time_order=2, space_order=2):
+        # Only dim=2 and dim=3 are supported
```
Comment (mlange05) [inline examples/tti_operators.py]: A proper docstring would be great.

**Comment by mlange05** [comment]
Excellent stuff, I particularly like the cleaned up examples. Good to go once minor style issues are addressed.

File: examples/fwi_operators.py (hunk 2 above)
Comment (navjotk) [inline examples/fwi_operators.py]: All I know is this class provides the coordinate transformations. I don't understand how it works. Is the "what?" enough for the docstring without the "how?"?

Comment (mlange05) [inline examples/fwi_operators.py]: Yes, just a brief statement of "what" alongside the existing comments on the expected input and output arguments should be fine.

**Comment by mlange05** [comment]
The current travis failure seems to be a sporadic timeout of the new TTI example. Any chance we can decrease the test size further, or is this purely the symbolic pre-processing? Also, are we happy with the result of the TTI run now?

**Comment by mlange05** [comment]
OK, this looks good to go. Merging...

**Pull request merged**